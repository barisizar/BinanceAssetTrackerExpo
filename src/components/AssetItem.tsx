import React, {useState, useEffect, useRef} from 'react';
import {View, Text, StyleSheet, Image, Animated} from 'react-native';
import {LineChart} from 'react-native-svg-charts';
import * as shape from 'd3-shape';
import {subscribeToTickerWebSocket} from '../services/api';

interface AssetItemProps {
  asset: {
    symbol: string;
    name: string;
    price: number;
    priceChangePercent: number;
    logoUrl: string;
    sparklineData: number[];
  };
}

export const AssetItem: React.FC<AssetItemProps> = ({asset}) => {
  const [currentPrice, setCurrentPrice] = useState(asset?.price);
  const [priceChangePercent, setPriceChangePercent] = useState(
    asset.priceChangePercent,
  );
  const [priceChangeColor, setPriceChangeColor] = useState('#1A1A1A');
  const fadeAnim = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    const unsubscribe = subscribeToTickerWebSocket(asset?.symbol, data => {
      const newPrice = parseFloat(data.c);
      setCurrentPrice(prevPrice => {
        if (newPrice !== prevPrice) {
          setPriceChangeColor(newPrice > prevPrice ? '#00C087' : '#FF3B30');
          animatePriceChange();
        }
        return newPrice;
      });
      setPriceChangePercent(parseFloat(data.P));
    });

    return () => {
      if (unsubscribe) {
        unsubscribe();
      }
    };
  }, [asset.symbol]);

  const animatePriceChange = () => {
    Animated.sequence([
      Animated.timing(fadeAnim, {
        toValue: 0,
        duration: 300,
        useNativeDriver: true,
      }),
      Animated.timing(fadeAnim, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true,
      }),
    ]).start(() => {
      setPriceChangeColor('#1A1A1A');
    });
  };

  const chartColor = priceChangePercent >= 0 ? '#00C087' : '#FF3B30';
  const changeColor = priceChangePercent >= 0 ? '#00C087' : '#FF3B30';

  const validSparklineData =
    asset?.sparklineData?.filter(value => Number.isFinite(value)) || [];
  const renderChart = validSparklineData.length >= 2;

  const formattedPrice = new Intl.NumberFormat('en-IN', {
    style: 'currency',
    currency: 'INR',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(currentPrice);

  return (
    <View style={styles.container}>
      <View style={styles.leftContent}>
        <Image source={{uri: asset?.logoUrl}} style={styles.logo} />
        <View style={styles.nameContainer}>
          <Text style={styles.name}>{asset?.name}</Text>
          <Text style={styles.symbol}>{asset?.symbol}</Text>
        </View>
      </View>
      {renderChart && (
        <View style={styles.chartContainer}>
          <LineChart
            style={{height: 40, width: 60}}
            data={validSparklineData}
            svg={{stroke: chartColor, strokeWidth: 2}}
            contentInset={{top: 10, bottom: 10}}
            curve={shape.curveNatural}
          />
        </View>
      )}
      <View style={styles.rightContent}>
        <Animated.Text
          style={[
            styles.price,
            {
              color: priceChangeColor,
              opacity: fadeAnim,
            },
          ]}>
          {formattedPrice}
        </Animated.Text>
        <Text style={[styles.change, {color: changeColor}]}>
          {priceChangePercent >= 0 ? '+' : ''}
          {priceChangePercent.toFixed(2)}%
        </Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: 8,
    paddingVertical: 12,
    backgroundColor: '#F8F9FA',
    borderRadius: 8,
    marginBottom: 8,
    marginHorizontal: 6,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  leftContent: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 3,
  },
  logo: {
    width: 32,
    height: 32,
    borderRadius: 16,
    marginRight: 12,
  },
  nameContainer: {
    justifyContent: 'center',
  },
  name: {
    fontSize: 16,
    fontWeight: '600',
    color: '#1A1A1A',
  },
  symbol: {
    fontSize: 14,
    color: '#888',
  },
  chartContainer: {
    flex: 3,
    justifyContent: 'center',
    alignItems: 'center',
    marginLeft: 16,
  },
  rightContent: {
    alignItems: 'flex-end',
    flex: 3,
  },
  price: {
    fontSize: 16,
    fontWeight: '600',
  },
  change: {
    fontSize: 14,
  },
});
