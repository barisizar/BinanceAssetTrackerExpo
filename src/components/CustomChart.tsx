import React, {useState, useEffect, useRef, useMemo} from 'react';
import {
  View,
  ScrollView,
  Text,
  StyleSheet,
  Dimensions,
  PanResponder,
} from 'react-native';
import Svg, {Path, Circle, Text as SvgText, Line} from 'react-native-svg';
import { formatPriceInINR } from '../utils/formatPrice';

const {width: SCREEN_WIDTH} = Dimensions.get('window');
const CHART_HEIGHT = 250;
const CHART_WIDTH = SCREEN_WIDTH * 3;
const X_AXIS_HEIGHT = 30;
const MAX_X_LABELS = 12;
const Y_AXIS_PADDING = 0.1; // 10% padding

interface ChartData {
  labels: string[];
  datasets: {data: number[]}[];
}

interface CustomChartProps {
  data: ChartData;
  onSelectDataPoint: (label: string, value: number) => void;
}

const CustomChart: React.FC<CustomChartProps> = ({data, onSelectDataPoint}) => {
  const [selectedPoint, setSelectedPoint] = useState<{
    x: number;
    y: number;
    value: number;
    label: string;
  } | null>(null);
  const scrollViewRef = useRef<ScrollView>(null);

  const {minValue, maxValue, path, dataPoints} = useMemo(() => {
    if (!data.datasets[0].data.length) {
      return {minValue: 0, maxValue: 0, path: '', dataPoints: []};
    }

    const minValue = Math.min(...data.datasets[0].data);
    const maxValue = Math.max(...data.datasets[0].data);

    // add padding
    const range = maxValue - minValue;
    const paddedMin = minValue - range * Y_AXIS_PADDING;
    const paddedMax = maxValue + range * Y_AXIS_PADDING;

    const getY = (value: number) => {
      return (
        CHART_HEIGHT -
        ((value - paddedMin) / (paddedMax - paddedMin)) * CHART_HEIGHT
      );
    };

    const dataPoints = data.datasets[0].data.map((point, index) => ({
      x: (index / (data.datasets[0].data.length - 1)) * CHART_WIDTH,
      y: getY(point),
      value: point,
      label: data.labels[index],
    }));

    const path = dataPoints.reduce((acc, point, index) => {
      return `${acc} ${index === 0 ? 'M' : 'L'} ${point.x} ${point.y}`;
    }, '');

    return {minValue: paddedMin, maxValue: paddedMax, path, dataPoints};
  }, [data]);

  const visibleLabels = useMemo(() => {
    const totalPoints = dataPoints.length;
    let step;

    if (totalPoints <= MAX_X_LABELS) {
      step = 1;
    } else if (totalPoints <= MAX_X_LABELS * 2) {
      step = 2;
    } else {
      step = Math.ceil(totalPoints / MAX_X_LABELS);
    }

    return dataPoints.filter((_, index) => index % step === 0);
  }, [dataPoints]);

  const panResponder = PanResponder.create({
    onStartShouldSetPanResponder: () => true,
    onMoveShouldSetPanResponder: () => true,
    onPanResponderGrant: evt =>
      handleTouch(evt.nativeEvent.locationX, evt.nativeEvent.locationY),
    onPanResponderMove: evt =>
      handleTouch(evt.nativeEvent.locationX, evt.nativeEvent.locationY),
  });

  const handleTouch = (touchX: number, touchY: number) => {
    const scrollX = scrollViewRef.current?.contentOffset?.x || 0;
    const adjustedX = touchX + scrollX;

    const closestPoint = dataPoints.reduce((closest, point) => {
      return Math.abs(point.x - adjustedX) < Math.abs(closest.x - adjustedX)
        ? point
        : closest;
    });

    setSelectedPoint(closestPoint);
    onSelectDataPoint(closestPoint.label, closestPoint.value);
  };

  useEffect(() => {
    setTimeout(() => {
      scrollViewRef.current?.scrollToEnd({animated: false});
    }, 100);
  }, [data.labels]);

  if (
    !data ||
    !data.datasets ||
    data.datasets.length === 0 ||
    !data.datasets[0].data ||
    data.datasets[0].data.length === 0
  ) {
    return (
      <View style={styles.errorContainer}>
        <Text>No data available</Text>
      </View>
    );
  }

  return (
    <View testID="custom-chart" style={styles.container} >
      <ScrollView
        ref={scrollViewRef}
        horizontal
        showsHorizontalScrollIndicator={false}
        contentContainerStyle={styles.scrollViewContent}
        {...panResponder.panHandlers}>
        <Svg width={CHART_WIDTH} height={CHART_HEIGHT + X_AXIS_HEIGHT}>
          <Path d={path} fill="none" stroke="#4A90E2" strokeWidth="2" />
          {selectedPoint && (
            <>
              <Circle
                cx={selectedPoint.x}
                cy={selectedPoint.y}
                r="5"
                fill="#4A90E2"
              />
              <SvgText
                x={selectedPoint.x}
                y={selectedPoint.y - 10}
                fill="black"
                fontSize="12"
                fontWeight="bold"
                textAnchor="middle">
                {formatPriceInINR(selectedPoint.value)}
              </SvgText>
            </>
          )}
          {/* X-axis line */}
          <Line
            x1="0"
            y1={CHART_HEIGHT}
            x2={CHART_WIDTH}
            y2={CHART_HEIGHT}
            stroke="#E0E0E0"
            strokeWidth="1"
          />
          {/* X-axis labels and ticks */}
          {visibleLabels.map((point, index) => (
            <React.Fragment key={index}>
              <Line
                x1={point.x}
                y1={CHART_HEIGHT}
                x2={point.x}
                y2={CHART_HEIGHT + 5}
                stroke="#E0E0E0"
                strokeWidth="1"
              />
              <SvgText
                x={point.x}
                y={CHART_HEIGHT + 20}
                fontSize="10"
                fill="#666"
                textAnchor="middle">
                {point.label}
              </SvgText>
            </React.Fragment>
          ))}
        </Svg>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
    backgroundColor: 'white',
  },
  scrollViewContent: {
    width: CHART_WIDTH,
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
  },
  errorContainer: {
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
  },
});

export default CustomChart;
