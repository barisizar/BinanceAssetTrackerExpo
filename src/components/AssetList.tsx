import React, {useCallback} from 'react';
import {
  FlatList,
  ActivityIndicator,
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {AssetItem} from './AssetItem';
import {useAssetData} from '../hooks/useAssetData';

export const AssetList: React.FC = () => {
  const {assets, loading, loadingMore, hasMore, loadMoreAssets, error} =
    useAssetData();
  const navigation = useNavigation();

  const handleAssetPress = useCallback(
    (asset: any) => {
      navigation.navigate('AssetDetail' as never, {asset} as never);
    },
    [navigation],
  );

  const renderItem = useCallback(
    ({item}: {item: any}) => (
      <TouchableOpacity onPress={() => handleAssetPress(item)}>
        <AssetItem asset={item} />
      </TouchableOpacity>
    ),
    [handleAssetPress],
  );

  if (loading && assets.length === 0) {
    return (
      <View style={styles.centered}>
        <ActivityIndicator size="large" testID="loading-indicator"/>
      </View>
    );
  }

  if (error && assets.length === 0) {
    return (
      <View style={styles.centered}>
        <Text>{error}</Text>
      </View>
    );
  }

  return (
    <View style={styles.container} testID="custom-chart">
      <FlatList
        data={assets}
        renderItem={renderItem}
        keyExtractor={item => item.symbol}
        onEndReachedThreshold={0.5}
        onEndReached={loadMoreAssets}
        ListFooterComponent={loadingMore ? <ActivityIndicator /> : null}
        contentContainerStyle={styles.listContent}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F8F9FA',
  },
  centered: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  header: {
    fontSize: 28,
    fontWeight: 'bold',
    padding: 16,
    paddingBottom: 8,
  },
  listContent: {
    paddingTop: 8,
    paddingBottom: 16,
  },
});
