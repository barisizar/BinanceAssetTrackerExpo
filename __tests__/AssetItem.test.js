import React from 'react';
import renderer from 'react-test-renderer';
import { AssetItem } from '../src/components/AssetItem';

jest.mock('../src/services/api', () => ({
    subscribeToTickerWebSocket: jest.fn(() => jest.fn()),
}));

jest.mock('react-native-svg-charts', () => ({
    LineChart: 'LineChart',
}));

describe('AssetItem', () => {
    const mockAsset = {
        symbol: 'BTC',
        name: 'Bitcoin',
        price: 50000,
        priceChangePercent: 5.0,
        logoUrl: 'https://example.com/btc.png',
        sparklineData: [49000, 50000, 51000, 50500, 50000],
    };

    it('renders correctly', () => {
        const tree = renderer.create(<AssetItem asset={mockAsset} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});