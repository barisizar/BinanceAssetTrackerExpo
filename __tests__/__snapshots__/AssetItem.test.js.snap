// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetItem renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8F9FA",
      "borderRadius": 8,
      "elevation": 3,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 8,
      "marginHorizontal": 6,
      "paddingHorizontal": 8,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 3,
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://example.com/btc.png",
        }
      }
      style={
        {
          "borderRadius": 16,
          "height": 32,
          "marginRight": 12,
          "width": 32,
        }
      }
    />
    <View
      style={
        {
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#1A1A1A",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        Bitcoin
      </Text>
      <Text
        style={
          {
            "color": "#888",
            "fontSize": 14,
          }
        }
      >
        BTC
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 3,
        "justifyContent": "center",
        "marginLeft": 20,
      }
    }
  >
    <LineChart
      contentInset={
        {
          "bottom": 10,
          "top": 10,
        }
      }
      curve={[Function]}
      data={
        [
          49000,
          50000,
          51000,
          50500,
          50000,
        ]
      }
      style={
        {
          "height": 40,
          "width": 60,
        }
      }
      svg={
        {
          "stroke": "#00C087",
          "strokeWidth": 2,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flex": 3,
      }
    }
  >
    <Text
      collapsable={false}
      style={
        {
          "color": "#1A1A1A",
          "fontSize": 16,
          "fontWeight": "600",
          "opacity": 1,
        }
      }
    >
      â‚¹50,000.00
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
          },
          {
            "color": "#00C087",
          },
        ]
      }
    >
      +
      5.00
      %
    </Text>
  </View>
</View>
`;
